Initial version: SQLite UDF API

tiagokepe opened the pull request:
This PR is a proof of "concept" for the support of SQLite UDF API #1792.

For now, it is just working with the _int32_ type, but when all relevant points are defined (e.g., casting back and forth the _sqlite_values_), other types can be support straightforward.

Refinements and much work are still needed, and suggestions/improvements are welcome.

tiagokepe committed:
Initial implementation of sqlite3_create_function().

tiagokepe committed:
First version of SQLite UDF API.

tiagokepe committed:
Created test directory.

tiagokepe committed:
Making CMake works.

tiagokepe committed:
Removed commented lines.

tiagokepe committed:
Renamed UDF.

tiagokepe committed:
Changed UDF create function to support ANY input type.

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_udf_api/include/udf_struct_sqlite3.h:
```diff
@@ -0,0 +1,29 @@
+        const char *zPType; /* Pointer type when MEM_Term|MEM_Subtype|MEM_Null */
+    } u;
+    // uint16_t flags;
+    PhysicalType type;
```
Rather than using the physical type here, we can use a different enum. Note that the types that need to be stored in here are rather limited. If we look at the [sqlite3_value_type function](http://sqlite.org/c3ref/value_blob.html), we see the following types:

```c
SQLITE_INTEGER, SQLITE_FLOAT, SQLITE_TEXT, SQLITE_BLOB, or SQLITE_NULL.
```

This corresponds with the methods used for extracting the actual data value from the sqlite3_value wrapper:

```c
const void *sqlite3_value_blob(sqlite3_value*);
double sqlite3_value_double(sqlite3_value*);
int sqlite3_value_int(sqlite3_value*);
sqlite3_int64 sqlite3_value_int64(sqlite3_value*);
void *sqlite3_value_pointer(sqlite3_value*, const char*);
const unsigned char *sqlite3_value_text(sqlite3_value*);
```

In our sqlite3_value, I would store the same types using the same type enum (`int type`).

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_udf_api/include/udf_struct_sqlite3.h:
```diff
@@ -0,0 +1,29 @@
+};
+
+struct sqlite3_context {
+    sqlite3_value result;
```
The list of functions that can operate on the sqlite3_context is [here](https://sqlite.org/c3ref/context.html). We will need to implement all of these, although some of them can be stubs for now (e.g. `sqlite3_aggregate_context` is a stub).

Of interest for the contents of the `sqlite3_context` are:
* [sqlite3_set/getauxdata](https://sqlite.org/c3ref/get_auxdata.html)
* [sqlite3_result_error](https://sqlite.org/c3ref/result_blob.html)
* [sqlite3_user_data](https://sqlite.org/c3ref/user_data.html)
* [sqlite3_context_db_handle](https://sqlite.org/c3ref/context_db_handle.html)

All of the necessary information to implement these functions should be stored within the context.

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_udf_api/sqlite3_udf_wrapper.cpp:
```diff
@@ -0,0 +1,42 @@
+			sqlite_udf(&context, argc, &value);
+
+			//casting the result sqlite value to string_t
+			if(TypeIsInteger(context.result.type)) {
```
It would be faster to do the result conversion in a vectorized manner as well, i.e. store all of the resulting values in an array and then bulk convert them in a loop.

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_udf_api/sqlite3_udf_wrapper.cpp:
```diff
@@ -0,0 +1,42 @@
+		args.data[0].Orrify(args.size(), vec_data);
+
+		vector<sqlite3_value> vec_sqlite(args.size());
+		CastSQLite::ToVector(args.data[0].GetType(), vec_data, vec_sqlite);
```
I think it should be possible to create a single function that handles an arbitrary number of arguments, rather than needing to create different functions for 1 argument, 2 arguments, 3 arguments, etc as we did in the vectorized UDFs.

The reason we did this in the vectorized UDFs is because the template specialisation on the number of arguments allows the function call to be inlined. This is significantly faster, because we save the function call per value.

For the SQLite UDFs, we unfortunately have no choice. The format of the function (a function pointer) does not allow us to inline it, and hence specialising on the number of arguments does not get us the same performance boost. 

Instead, you could create a single function that handles all the argument types here, which would perform a loop over the input vectors (`args.ColumnCount()` and constructs the `sqlite3_value` arguments for each of the function invocations.

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:
```diff
@@ -1120,8 +1123,17 @@ int sqlite3_create_function(sqlite3 *db, const char *zFunctionName, int nArg, in
+	string fname = string(zFunctionName);
+
+	//Unary function for now
+	if(nArg == 1 && xFunc) {
```
See above, I think the specialisation on `nArg` can be avoided.

Mytherin reviewed (commented):
Thanks for the work! This is definitely on the right track to what we are looking for. Some comments:

tiagokepe committed:
Created enum SQLiteTypeValue.

tiagokepe committed:
Added suport for all integer and float types.

tiagokepe committed:
Result conversion in a vectorized fashion.

tiagokepe committed:
SQLite UDF API: handling an arbitrary number of arguments.

tiagokepe committed:
SQLite UDF API: Added support for many methods:
- sqlite3_value_* and sqlite3_result_* for integer, float, text, and blob types;
- get_user_data(), set_null, and value_type;
- traverse/cast arrays in a vectorized fashion;
- generic create udf function for any number of input arguments.

tiagokepe committed:
Run make format.

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:
```diff
@@ -1386,27 +1429,38 @@ SQLITE_API int sqlite3_result_zeroblob64(sqlite3_context *, sqlite3_uint64 n) {
-double sqlite3_value_double(sqlite3_value *) {
-	return 0;
+double sqlite3_value_double(sqlite3_value *pVal) {
+	if (pVal->type == SQLiteTypeValue::FLOAT) {
```
I believe we should try to cast the value in the case the underlying value is an integer or text value, see e.g. [the docs](https://sqlite.org/c3ref/value_blob.html):

> As long as the input parameter is correct, these routines can only fail if an out-of-memory error occurs during a format conversion. Only the following subset of interfaces are subject to out-of-memory errors:
sqlite3_value_blob()
sqlite3_value_text()
sqlite3_value_text16()
sqlite3_value_text16le()
sqlite3_value_text16be()
sqlite3_value_bytes()
sqlite3_value_bytes16()

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:
```diff
@@ -1386,27 +1429,38 @@ SQLITE_API int sqlite3_result_zeroblob64(sqlite3_context *, sqlite3_uint64 n) {
 
-sqlite3_int64 sqlite3_value_int64(sqlite3_value *) {
+sqlite3_int64 sqlite3_value_int64(sqlite3_value *pVal) {
+	if (pVal->type == SQLiteTypeValue::INTEGER) {
```
Same here, I believe we should add a cast here if the underlying value type is different.

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:
```diff
@@ -1386,27 +1429,38 @@ SQLITE_API int sqlite3_result_zeroblob64(sqlite3_context *, sqlite3_uint64 n) {
+	if (pVal && (pVal->type == SQLiteTypeValue::TEXT || pVal->type == SQLiteTypeValue::BLOB)) {
+		// We don't need to append \0 at the end of string
+		// To know the number of chars or bytes, call function sqlite3_value_bytes()
+		return (const unsigned char *)pVal->str_t.GetDataUnsafe();
```
Same here.

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_udf_api/include/sqlite3_value_type.hpp:
```diff
@@ -0,0 +1,3 @@
+#pragma once
+
+enum class SQLiteTypeValue : uint8_t { INTEGER = 1, FLOAT = 2, TEXT = 3, BLOB = 4, NULL_VALUE = 5 };
```
Could you wrap this in the duckdb namespace?

Mytherin commented on tools/sqlite3_api_wrapper/test/include/sqlite_stmt_wrapper.hpp:
```diff
@@ -0,0 +1,80 @@
+	 */
+	int ExecutePreparedStmt() {
+		if (!stmt) {
+			fprintf(stderr, "There is no prepated statement: Prepare(...) must be invoked firstly.\n");
```
prepared

Mytherin commented on tools/sqlite3_api_wrapper/test/test_sqlite3_udf_api_wrapper.cpp:
```diff
@@ -0,0 +1,186 @@
+	// create sqlite udf
+	REQUIRE(sqlite3_create_function(db_w.db, "multiply10", 1, 0, nullptr, &multiply10, nullptr, nullptr) == SQLITE_OK);
+	REQUIRE(db_w.Execute("SELECT multiply10(i) FROM integers"));
+	REQUIRE(db_w.CheckColumn(0, {"-50", "-40", "-30", "-20", "-10", "0", "10", "20", "30", "40", "50"}));
```
Could you add a test with NULL inputs?

Mytherin commented on tools/sqlite3_api_wrapper/test/test_sqlite3_udf_api_wrapper.cpp:
```diff
@@ -0,0 +1,186 @@
+	REQUIRE(sqlite3_create_function(db_w.db, "multiply10", 1, 0, nullptr, &multiply10, nullptr, nullptr) == SQLITE_OK);
+	REQUIRE(db_w.Execute("SELECT multiply10(i) FROM integers"));
+	REQUIRE(db_w.CheckColumn(0, {"-50", "-40", "-30", "-20", "-10", "0", "10", "20", "30", "40", "50"}));
+}
```
Could you add a test that results in an error (e.g. a division operator, that throws an error on zero on the rhs)?

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:
```diff
@@ -1323,16 +1343,26 @@ SQLITE_API char *sqlite3_win32_unicode_to_utf8(LPCWSTR zWideText) {
 // TODO complain
-SQLITE_API void sqlite3_result_blob(sqlite3_context *, const void *, int, void (*)(void *)) {
+SQLITE_API void sqlite3_result_blob(sqlite3_context *context, const void *blob, int n_bytes, void (*)(void *)) {
+	if (blob) {
```
We prefer to [return early](https://medium.com/swlh/return-early-pattern-3d18a41bba8):
```cpp
if (!blob) {
   return;
}
```

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:
```diff
@@ -1120,8 +1125,22 @@ int sqlite3_create_function(sqlite3 *db, const char *zFunctionName, int nArg, in
+	string fname = string(zFunctionName);
+
+	// Scalar function
+	if (xFunc) {
```
Same here: return early.

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:
```diff
@@ -1120,8 +1125,22 @@ int sqlite3_create_function(sqlite3 *db, const char *zFunctionName, int nArg, in
-	// fprintf(stderr, "sqlite3_create_function: unsupported.\n");
-	return -1;
+
+	string fname = string(zFunctionName);
```
What if `zFunctionName` is `nullptr`?

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:
```diff
@@ -1347,16 +1377,29 @@ SQLITE_API void sqlite3_result_error_nomem(sqlite3_context *) {
+	if (str_c) {
+		auto utf_type = Utf8Proc::Analyze(str_c, n_chars);
+		if (utf_type == UnicodeType::INVALID) {
+			throw Exception("Text value is not valid UTF8");
```
You should not throw exceptions from within a C API function. C programs cannot deal with exceptions as they are a C++ function, so this will always result in unpredictable behavior and/or aborting the program.

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_udf_api/cast_sqlite.cpp:
```diff
@@ -0,0 +1,137 @@
+		return ToVectorSQLiteValue<string_t>((string_t *)vec_data.data, size);
+	}
+	default:
+		throw std::runtime_error("SQLite UDF API: type is not supported!");
```
Perhaps as a fallback cast the input vector to `VARCHAR`?

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_udf_api/cast_sqlite.cpp:
```diff
@@ -0,0 +1,137 @@
+unique_ptr<vector<sqlite3_value>> CastSQLite::ToVectorSQLite(LogicalType type, VectorData &vec_data, idx_t size) {
+	LogicalTypeId type_id = type.id();
+	switch (type_id) {
+	case LogicalTypeId::TINYINT: {
```
LogicalType::NULL is missing. Could you add a test in the form of `SELECT my_sqlite_function(NULL)`?

Mytherin reviewed (commented):
Thanks for the updates! Looking good. Some comments:

tiagokepe committed:
Merge remote-tracking branch 'upstream/master'

tiagokepe committed:
Formatting missing changes.

tiagokepe committed:
SQLite UDF API, many changes:
- Casting unsupported logical types to VARCHAR;
- Casting sqlite_values;
- Testing NULL values;

tiagokepe committed:
Casting Timestamp types to VARCHAR.

tiagokepe committed:
Run format fix.

tiagokepe committed:
Formatting...

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_udf_api/cast_sqlite.cpp:
```diff
@@ -0,0 +1,210 @@
+			}
+
+			data_chunk.data[i].SetType(vchar_type);
+			VectorOperations::Copy(varchar_vector, data_chunk.data[i], data_chunk.size(), 0, 0);
```
Instead of both casting and copying, could we create a DataChunk that either (1) performs a cast if required, or (2) references the input vector of the other data chunk if not required? That would save a copy and simplify the code here somewhat.

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_udf_api/include/udf_struct_sqlite3.h:
```diff
@@ -0,0 +1,61 @@
+#include "sqlite3.h"
+#include "sqlite3_value_type.hpp"
+
+using namespace duckdb;
```
Please avoid `using namespace`

Mytherin commented on tools/sqlite3_api_wrapper/test/include/sqlite_db_wrapper.hpp:
```diff
@@ -0,0 +1,66 @@
+#include <vector>
+#include "util_functions.hpp"
+
+using namespace std;
```
Please avoid `using namespace`

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_udf_api/sqlite3_udf_wrapper.cpp:
```diff
@@ -0,0 +1,85 @@
+		// traversing the vector of sqlite values
+		for (idx_t row_idx = 0; row_idx < res_sqlite_values.size(); ++row_idx) {
+
+			sqlite3_value **argv = new sqlite3_value *[argc];
```
Could we avoid allocating this for every row iteration?
Could we wrap this in a unique_ptr, to avoid having to manually call delete later on (e.g. `unique_ptr<sqlite_value[]>(new sqlite_value[argc])`)?

Mytherin commented on tools/sqlite3_api_wrapper/sqlite3_udf_api/sqlite3_udf_wrapper.cpp:
```diff
@@ -0,0 +1,85 @@
+
+			// create a tuple from sqlite_values
+			for (idx_t col_idx = 0; col_idx < argc; ++col_idx) {
+				argv[col_idx] = &(*(vec_sqlite[col_idx]))[row_idx];
```
Perhaps you could directly construct the sqlite3_value objects in the correct row-wise format in `CastSQLite::ToVectorsSQLiteValue`? That avoids having to move them from column-wise format to row-wise format here.

tiagokepe replied:
Hey Mark,
Thanks for the suggestions, I already implemented them all.

But I guess this one is a little problematic. I don't think so directly construct the sqlite3_value objects in the ToVectorsSQLiteValue is a better option because we still need to perform the tuple reconstruction.
- Constructing the tuple in the CastSQLite::ToVector may inhibit the vectorized cast to sqlite3_values.
- Another issue is the values of the argv (**argv) are pointers, so we still need to store sqlite values in some place, and then constructing the argv argument to point to the sqlite3_values.

P.S. I've already implemented such code, if you want I can share with you.

Mytherin replied:
Hey Tiago, you're correct, the fact that they are pointers complicates it somewhat and the benefits of constructing them in row-wise format in the `ToVectorsSQLiteValue` is significantly lower as we would still need to set up the pointers. I'm fine with leaving this as-is.

Could you update the PR with your new code?

tiagokepe replied:
I pushed the code and the workflow was triggered on my git fork.
But it appears that the workflow on duckdb repository was not triggered automatically.

Mytherin reviewed (commented):
Thanks for the updates! Looking good. Some more feedback:

tiagokepe committed:
SQLite UDF API: several improvements.

tiagokepe committed:
SQLite UDF API: testing cast operations.

tiagokepe committed:
Reorganized namespace usage.

tiagokepe committed:
Using duckdb numerics to prevent Windows' macro (min and max) expansion.

Mytherin commented:
Could you have a look at resolving the merge conflicts with master?

tiagokepe committed:
Merge with DuckDB master repo.

tiagokepe committed:
Fix merge conflicts.

tiagokepe committed:
Run format.

tiagokepe committed:
Merge remote-tracking branch 'upstream/master'

tiagokepe committed:
SQLite UDF API: supporting functions with variadic arguments.

tiagokepe committed:
CreateFunction: supporting function overload.

tiagokepe committed:
Disabled test with SQL syntax error.

tiagokepe committed:
Run format.

tiagokepe committed:
Small fixes.

tiagokepe committed:
SQLite UDF API: testing function overload.

Mytherin commented:
Thanks, looks great now!
